File: CrackingTheCodingInterview/Udemy/WaterContainers.java
package CrackingTheCodingInterview.Udemy;

public class WaterContainers {
    public static void main(String[] args) {
        
    }

    public int maxArea(int[] height) {
        int maxArea = Math.min(height[0], height[height.length-1])*height.length-1;
        int i = 0;
        int j = height[height.length-1];
        while (i < j) {
            if (height[i+1] > height[j-1]) {
                i++;
            } else {
                j++;
            }
            maxArea = Math.max(maxArea, Math.min(height[j], height[i])*(j-i));
        }
        return maxArea;
    }
}

File: CrackingTheCodingInterview/Udemy/WaterContainers.js
const waterContainers = function(heights) {
    let p1 = 0
    let p2 = heights.length-1
    let maxArea = 0;

    while (p1 < p2) {
        const height = Math.min(heights[p1], heights[p2]);
        const width = p2- p1;
        const area = height * width;
        maxArea = Math.max(maxArea,area);

        if (heights[p1] <= heights[p2]) {
            p1++
        } else {
            p2--;
        }

    }
    return maxArea;
}

console.log(waterContainers([1,8,6,2,9,4]))